---
output:
  md_document:
    variant: gfm
---

### Using `gen.simuHaplo`
This example will show how to use `gen.simuHaplo` to drop uniquely labelled founder genotypes down a genealogy to create proband haplotypes.  

First load `GENLIB`

```{r, warning=FALSE, message=FALSE}
library(GENLIB)
```
For this example we will use the example genealogy "genea140" that comes with GENLIB.
```{r, warning=FALSE}
data(genea140)
head(genea140)
```
The genealogy file must use integer ID's, and contain the 'ind', 'mother', 'father' columns.   
Use the `gen.genealogy` function to create a genealogy object
```{r warning=FALSE}
gen_obj <- gen.genealogy(genea140)
gen_obj
```

Now we can simulate transmission of haplotypes from founders to probands using `gen.simuhaplo`. The function can only simulate one chromosome at a time, so for multiple chromosomes will require multiple function calls (as different chromosomes will have different length, genetic length, and recombination patterns). Note that multiple function calls will over-write the previous output files, so consider using the `dir.create` function or something similar in a loop so that the output is written to different folders. For simulating sub-chromosomal regions, treat the region as a 'mini chromosome' and use appropriate parameters. The function takes the following parameters:  

**gen** : genealogy object   
**pro** : vector of proband ID's to simulate. If not specified will default to all probands in the specified genealogy.   
**ancestors** : Vector of founder ID's to include. If not specified will default to all founders of the specified probands. Founders in the genealogy that are not included in this vector will have their haplotypes represented by "0".      
**simulNo** : Number of simulations to run.    
**model** : Integer used to specify the model used to simulate meiosis. 1 = Poisson model. 2 = Zero-truncated Poisson model. 3 = Gamma renewal process. Both the zero-truncated Poisson, and Gamma models are not appropriate for sub-chromosomal regions.  
**model_params** : All the models require 1 parameter. **model_params** should be a vector of length 2, where the first element specifies the parameter for males, and the second for females. For details on the model parameters see the paper (cite).  
**cM_len** : A vector of length 2. Length of the chromosome to be simulated in centimorgans. The first element should be the length for males, and the second for females.  
**BP_len** : Integer value for the length of the chromosome in base pairs. Only a single value as it is the same for males and females.  
**physical_map_female** : Option to specify the relationship between genetic length and physical length. A dataframe with required columns named "BP" and "cM", where each row is a pair of points, and the map is constructed by linear interpolation of the specified points. The first row should be (0,0), and the last row (BP_len, cM_len). If not specified the map will be assumed to be linear across the chromosome.  
**physical_map_male** : Same as above but for males.  
**seed** : can seed the random number generator used, for replication purposes.  
**all_nodes** : Either 1 or 0. If 1 the function will produce an additional output file for the haplotypes of all the internal individuals of the genealogy (not just the probands).  
**outDir** : Directory to write output files to. Default is the current working directory.   

For this example we will run 100 simulations on the above genealogy. We will simulate a hypothetical chromosome 1 with a length of 290,000,000 BP. We use the Poisson process model of meiosis, so the parameters are the expected number of recombinations. In this case the model parameters are the same as the Morgan length of the region we are simulating, but for other models of meiosis this is not the case so we need to specify the genetic length of the region separately (`cM_len` parameter).
```{r cache=TRUE, message=FALSE}
#print processor specs for the benchmark
specs <- read.csv(text = system("systeminfo /FO csv", intern = TRUE))
print(specs$Processor.s.)

start_time <- Sys.time() #get speed of simulation

gen.simuHaplo(gen_obj, model = 1, simulNo = 100, model_params = c(1.98, 3.28), 
              cM_len = c(198, 328), BP_len = 290000000, all_nodes = 1)
end_time <- Sys.time()

print(end_time - start_time)
```
  
The function will create the file "Proband_Haplotypes.txt" in the specified directory. If the **all_nodes** parameter were set to 1 then it would also create the file "All_nodes_haplotypes.txt".  

### Output files    
First we will look at the structure of the "Proband_Haplotypes.txt" output file:
```{r}
f <- file(paste(getwd(),"Proband_Haplotypes.txt",sep="/"),"r")
print(readLines(f,n=3))
```
The first line of the output is the number of simulations, and the number of probands. In the rest of the file there will be one line for every proband for every simulation. So there will be num_simul * num_proband lines in the file. Each line contains information contained in a set of 3 curly braces. In the first curly bracket is the simulation number, the proband ID, and a placeholder spot. The following two sets of brackets contain the haplotype information for the diploid chromosome. The first of these two is the paternal chromosome copy, and the second is the maternal.
  
  The haplotypes are encoded **{position ; origin_ID; position; origin_ID; ... origin_ID; position}** where the positions are BP positions along the chromosome, and origin_ID refers to the ID of the founder from which the segment between those positions originated. The origin_ID are founder ID numbers with ".1" or ".2" appended to them, to distinguish between the diploid copies of the founder chromosome. 
  
The **"All_nodes_haplotypes.txt"** file is formatted in a similar way. But the first set of brackets contains information on the recombination positions. The next two brackets have the same format as the other file.
```{r}
f <- file(paste(getwd(),"All_nodes_haplotypes.txt",sep="/"),"r")
print(readLines(f,n=3))
```
The first line is the number of simulations, the number of probands, and the number of individuals. The number of individuals does not include the founders because they are not output in the "All_nodes_haplotypes.txt" file (because the founder haplotypes are trivial, they are all assumed unique).  

In the first set of curly braces is the recombination history. This is used internally to trace back upwards the simulation after the fact. In these brackets is four fields separated by semicolons. The first field is the simulation number, the second is the individual ID, the third and fourth contain the recombination history. These third and fourth fields are both formatted the same way, to represent the recombinations in the paternal, and maternal copies of the chromosome respectively.  

The recombination history is a sequence of numbers separated by commas. The **first** number is the "start" of the chromosome. So for example, if we are looking at the recombination history of the paternal copy of the chromosome, then if this first value is 0 that means the paternal chromosome starts with the father's paternal copy, if it is 1 then it starts with the fathers maternal copy. When recombinations are simulated, they are done by creating a new haplotype by stitching together the two copies of that parent's chromosome, swapping at every recombination position. But this operation can start with either of the parental copy, chosen randomly. The **second** value is the number of recombinations that occured during that meiosis. The **third** and following values are the positions of all the recombinations.